### PR TITLE
[CSM-378] Move tx fees logic to `Pos.Client.Txp`

### PR DESCRIPTION
**Pull request by flyingleafe**
Transaction creation logic has been simplified, and tx fees logic (together with boot distribution logic) was moved inside `create*Tx` functions.

There's some ugly decisions in this PR which made it possible to change logic without profound refactoring. Though I find these decisions ugly, I'd better not include such refactoring in this particular PR and implement it separately - because we need to start testing this logic (including making unit tests, CSM-399) asap.

Also, changing fees constants is not present in this PR, because it's present in #1269.

This PR has WIP label because it needs to be tested properly before merging (including Daedalus acceptance tests). PR is made in advance to give more time for review.

### CODE REVIEW

File: src/Pos/Client/Txp/Addresses.hs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+       ) where
+
+import           Universum
+
+import           Pos.Types (Address)
+
+-- | A class which have the method to generate a new address
+class Monad m => MonadAddresses m where
+    type AddrData m :: *
+    getNewAddress :: AddrData m -> m Address
```
Comment (flyingleafe) [inline src/Pos/Client/Txp/Addresses.hs]: I really don't like this. I'd rather have `getNewAddress :: m Address`, but the only way for implementing such instance for real wallet which I could think of is having some `newtype UglyWrapper m = UglyWrapper (ReaderT addrData m) deriving (<thousands of monads>)` and implementing an instance for it. Though we used to do such things, now it's a no-way, as I understand. Is there a better way to implement such a typeclass?

File: src/Pos/Wallet/Web/Server/Methods.hs [hunk 2]
```diff
@@ -586,27 +548,20 @@ getMoneySourceWallet (AddressMoneySource addrId) = cwamWId addrId
...
-        InternalError . sformat ("UnknownFeePolicy, tag: "%build)
+getMoneySourceUtxo :: WalletWebMode m => MoneySource -> m Utxo
+getMoneySourceUtxo =
+    getMoneySourceAddresses >=>
+    mapM (decodeCIdOrFail . cwamWId) >=>
+    getOwnUtxos
+
+-- [CSM-237] It should be moved to `Pos.Wallet.Web.Mode`, but
+-- to make it possible all this mess should be neatly separated
+-- to modules and refactored
```
Comment (flyingleafe) [inline src/Pos/Wallet/Web/Server/Methods.hs]: P.S: it's because of imports which are really hard to be made non-circular in this situation.

File: src/Pos/Client/Txp/Util.hs [hunk 3]
```diff
@@ -206,60 +260,184 @@ prepareInpsOuts utxo addrs outputs = do
...
+        -- Fee depends on size of tx in bytes, sign of a tx has the fixed size
+        -- so we can use arbitrary signer.
+        srcAddrs = NE.map fst trInputs
+        (_, fakeSK) = deterministicKeyGen "patakbardaqskovoroda228pva1488kk"
+        hdwSigners = NE.zip (NE.repeat $ fakeSigner fakeSK) srcAddrs
+    in makeMPubKeyTxAddrs hdwSigners trInputs txOutsWithRem
+
+    -- buildDistribution :: TxRaw -> Text
+    -- buildDistribution TxRaw{..} =
+    --     let entries =
```
Comment (flyingleafe) [inline src/Pos/Client/Txp/Util.hs]: This function was used only for logging in the old code. Now the code, identical to old one, is run in environment without `WithLogger`. Is it better to add `WithLogger` to `TxDistrMode` or can we simply abandon this log line?

File: lwallet/Main.hs [hunk 4]
```diff
@@ -217,9 +218,10 @@ runCmd sendActions (SendToAllGenesis duration conc delay_ sendMode tpsSentFile)
                 | otherwise = (atomically $ tryReadTQueue txQueue) >>= \case
                       Just (key, txOut, neighbours) -> do
                           utxo <- getOwnUtxo $ makePubKeyAddress $ safeToPublic (fakeSigner key)
-                          tx <- createTx utxo (fakeSigner key) (one (TxOutAux txOut []))
+                          tx <- createTx utxo (fakeSigner key) (one (TxOutAux txOut [])) (makePubKeyAddress $ toPublic key)
                           case tx of
-                              Left err -> addTxFailed tpsMVar >> logError (sformat ("Error: "%stext%" while trying to send to "%shown) err neighbours)
+                              Left (TxError err) -> addTxFailed tpsMVar >>
+                                  logError (sformat ("Error: "%stext%" while trying to send to "%shown) err neighbours)
```
Comment (Martoon-00) [inline lwallet/Main.hs]: I think it will look much nicer within `do` block

File: src/Pos/Client/Txp/Util.hs [hunk 5]
```diff
@@ -174,27 +220,35 @@ makeRedemptionTx rsk txInputs = makeAbstractTx mkWit (map ((), ) txInputs)
...
+        evalStateT (pickInputs []) (totalMoneyWithFee, sortedUnspent)
     case nonEmpty futxo of
-        Nothing       -> throwError "Failed to prepare inputs!"
-        Just inputsNE -> pure (map formTxInputs inputsNE, outputs)
+        Nothing       -> throwTxError "Failed to prepare inputs!"
+        Just inputsNE -> do
+            let totalTxAmount = unsafeSumTxOuts $ map snd inputsNE
+                trInputs = map formTxInputs inputsNE
+                trOutputs = outputs
+                trRemaining = totalTxAmount `unsafeSubCoin` totalMoneyWithFee
```
Comment (Martoon-00) [inline src/Pos/Client/Txp/Util.hs]: Should fee go to to change address?

File: src/Pos/Client/Txp/Util.hs [hunk 6]
```diff
@@ -206,60 +260,184 @@ prepareInpsOuts utxo addrs outputs = do
...
-    runExceptT $ do
-        properOutputs <- overrideTxDistrBoot outputs
-        uncurry (makePubKeyTx ss) <$>
-            prepareInpOuts
-                utxo
-                (makePubKeyAddress $ safeToPublic ss)
-                properOutputs
+    -> AddrData m
+    -> m (Either TxError TxAux)
+createTx utxo ss outputs addrData = runExceptT $ do
```
Comment (Martoon-00) [inline src/Pos/Client/Txp/Util.hs]: `createTx`, `createMTx` and some functions below got almost identical, it's worth extracting some logic

File: src/Pos/Client/Txp/Util.hs [hunk 7]
```diff
@@ -206,60 +260,184 @@ prepareInpsOuts utxo addrs outputs = do
...
+                 createFakeTxFromRawTx txRaw
+        if expectedFee == txFee
+            then pure txRaw
+            else stabilizeTxFeeDo (attempt - 1) txFee
+
+txToLinearFee
+    :: MonadError TxError m
+    => TxSizeLinear -> TxAux -> m TxFee
+txToLinearFee linearPolicy =
+    either throwError pure .
```
Comment (Martoon-00) [inline src/Pos/Client/Txp/Util.hs]: You just added that cool `eitherToThrow` :crying_cat_face:

File: src/Pos/Client/Txp/Util.hs [hunk 8]
```diff
@@ -206,60 +260,184 @@ prepareInpsOuts utxo addrs outputs = do
...
+  where
+    invalidFee reason = TxError ("Invalid fee: " <> reason)
+
+createFakeTxFromRawTx :: TxRaw -> TxAux
+createFakeTxFromRawTx TxRaw{..} =
+    let fakeAddr = txOutAddress . toaOut . NE.head $ trOutputs
+        fakeOutMB
+            | trRemaining == mkCoin 0 = Nothing
+            | otherwise = Just $ TxOutAux (TxOut fakeAddr trRemaining) []
+        txOutsWithRem = maybe trOutputs (\remTx -> remTx :| toList trOutputs) fakeOutMB
```
Comment (Martoon-00) [inline src/Pos/Client/Txp/Util.hs]: Also nice place for your new function

File: src/Pos/Client/Txp/Util.hs (hunk 3 above)
Comment (Martoon-00) [inline src/Pos/Client/Txp/Util.hs]: I think that `WithLogger` should be supposed by any ~large constraint we use, so if this is the only obstacle then why not to use this function.
On the other hand, this case seems too minor to log, it may be even better not to litter logs with this :man_shrugging:

File: src/Pos/Wallet/Web/Tracking.hs [hunk 9]
```diff
@@ -578,19 +574,3 @@ decryptAccount (hdPass, wCId) addr@(PubKeyAddress _ (Attributes (AddrPkAttrs (Ju
     guard $ length derPath == 2
     pure $ CWAddressMeta wCId (derPath !! 0) (derPath !! 1) (addressToCId addr)
 decryptAccount _ _ = Nothing
-
--- TODO [CSM-237] Move to somewhere (duplicate getWalletsAddrMetas from Methods.hs)
-getWalletAddrMetasDB
```
Comment (Martoon-00) [inline src/Pos/Wallet/Web/Tracking.hs]: Woow :fist_raised:

File: src/Pos/Wallet/Web/Server/Methods.hs [hunk 10]
```diff
@@ -586,27 +548,20 @@ getMoneySourceWallet (AddressMoneySource addrId) = cwamWId addrId
...
-                calculateTxSizeLinear linearPolicy .
-                biSize @TxAux $ txAux
-  where
-    invalidFee reason = InternalError ("Invalid fee: " <> reason)
-    unknownFeePolicy =
-        InternalError . sformat ("UnknownFeePolicy, tag: "%build)
+getMoneySourceUtxo :: WalletWebMode m => MoneySource -> m Utxo
+getMoneySourceUtxo =
+    getMoneySourceAddresses >=>
+    mapM (decodeCIdOrFail . cwamWId) >=>
```
Comment (Martoon-00) [inline src/Pos/Wallet/Web/Server/Methods.hs]: You meant `cwamId`?

File: src/Pos/Wallet/Web/Server/Methods.hs [hunk 11]
```diff
@@ -615,235 +570,47 @@ sendMoney
...
+                       _txOutputs tx
+            srcWallet = getMoneySourceWallet moneySource
+        logInfo $
+            sformat ("Successfully spent money from "%
+                     listF ", " addressF % " addresses on " %
+                     listF ", " addressF)
+            (toList srcAddrs)
+            dstAddrs
+
+        -- TODO: this should be removed in production
```
Comment (Martoon-00) [inline src/Pos/Wallet/Web/Server/Methods.hs]: Ahaha, I wonder how long these todos are going to hang there further

File: src/Pos/Wallet/Web/Server/Methods.hs [hunk 12]
```diff
@@ -615,235 +570,47 @@ sendMoney
...
+    srcAddrs <- forM addrMetas $ decodeCIdOrFail . cwamId
+
+    withSafeSigners passphrase sks $ \ss -> do
+        let hdwSigner = NE.zip ss srcAddrs
+        relatedAccount <- getSomeMoneySourceAccount moneySource
+        outputs <- coinDistrToOutputs dstDistr
+        TxAux {taTx = tx} <- rewrapTxError "Cannot send transaction" $
+            submitMTx enqueueMsg hdwSigner outputs (relatedAccount, passphrase)
+
+        ownUtxo <- getOwnUtxos $ toList srcAddrs
```
Comment (Martoon-00) [inline src/Pos/Wallet/Web/Server/Methods.hs]: Aren't you trying to find just spent `TxIn`s in utxo?

File: lwallet/Main.hs [hunk 13]
```diff
@@ -145,20 +145,21 @@ addTxSubmit mvar = modifySharedAtomic mvar (\(TxCount submitted failed sending)
...
-    etx <-
-        withSafeSigner (skeys !! idx) (pure emptyPassphrase) $ \mss ->
-        runEitherT $ do
-            ss <- mss `whenNothing` throwError "Invalid passphrase"
-            lift $ submitTx
-                (immediateConcurrentConversations sendActions na)
-                ss
-                (map (flip TxOutAux []) outputs)
+    let skey = skeys !! idx
+        curAddr = makePubKeyAddress $ encToPublic skey
```
Comment (Martoon-00) [inline lwallet/Main.hs]: Ah, here we use root key as address directly, so nevermind.
Unless heavy and light wallets share the same key storage - then everything is sad, and it is for a long time already

File: src/Pos/Client/Txp/Util.hs (hunk 5 above)
Comment (flyingleafe) [inline src/Pos/Client/Txp/Util.hs]: @Martoon-00 Fee is not going anywhere, it just gets lost (if it will go to a change address, it wouldn't be _spent_, and it wouldn't effectively be a _fee_

File: src/Pos/Client/Txp/Util.hs (hunk 7 above)
Comment (flyingleafe) [inline src/Pos/Client/Txp/Util.hs]: Read again: here we don't do `throwM`, but `throwError`.

File: src/Pos/Client/Txp/Util.hs (hunk 8 above)
Comment (flyingleafe) [inline src/Pos/Client/Txp/Util.hs]: which function? you mean `mkOutputsWithRem`?
but it's monadic, and here we have pure context. Though it may be a good idea to abstract a pure part away.

File: src/Pos/Wallet/Web/Server/Methods.hs (hunk 10 above)
Comment (flyingleafe) [inline src/Pos/Wallet/Web/Server/Methods.hs]: Hmm, yeah, I think you're right (gosh, these fields look so identical, it's really error-prone)

File: src/Pos/Wallet/Web/Server/Methods.hs (hunk 12 above)
Comment (flyingleafe) [inline src/Pos/Wallet/Web/Server/Methods.hs]: Here I'm trying to find `TxOut`s corresponding to spent `TxIn`s, which are required for making `THEntry`

File: src/Pos/Wallet/Web/Server/Methods.hs (hunk 11 above)
Comment (flyingleafe) [inline src/Pos/Wallet/Web/Server/Methods.hs]: Tbh I don't even remember what exactly it's about...

File: lwallet/Main.hs (hunk 13 above)
Comment (flyingleafe) [inline lwallet/Main.hs]: Yeah, if we need to migrate light wallet to hd addresses then much more job should be done, I suppose.

File: src/Pos/Client/Txp/Util.hs (hunk 5 above)
Comment (Martoon-00) [inline src/Pos/Client/Txp/Util.hs]: Aha, I see now

File: src/Pos/Client/Txp/Util.hs (hunk 7 above)
Comment (Martoon-00) [inline src/Pos/Client/Txp/Util.hs]: Ah, rly

File: src/Pos/Wallet/Web/Server/Methods.hs (hunk 12 above)
Comment (Martoon-00) [inline src/Pos/Wallet/Web/Server/Methods.hs]: Why do you do it just after you created transaction which seem to remove inputs you seek?

File: src/Pos/Client/Txp/Util.hs [hunk 14]
```diff
@@ -206,60 +261,208 @@ prepareInpsOuts utxo addrs outputs = do
...
+-----------------------------------------------------------------------------
+
+prepareTxWithFee
+    :: TxDistrMode ctx m
+    => Utxo
+    -> TxOutputs
+    -> ExceptT TxError m TxRaw
+prepareTxWithFee utxo outputs = do
+    feePolicy <- bvdTxFeePolicy <$> lift gsAdoptedBVData
+    case feePolicy of
```
Comment (volhovM) [inline src/Pos/Client/Txp/Util.hs]: I advice to use lambdacase here.

File: src/Pos/Client/Txp/Util.hs [hunk 15]
```diff
@@ -206,60 +261,208 @@ prepareInpsOuts utxo addrs outputs = do
...
+
+-- Read-only function.
+computeTxFee
+    :: TxDistrMode ctx m
+    => Utxo
+    -> TxOutputs
+    -> ExceptT TxError m TxFee
+computeTxFee utxo outputs = do
+    feePolicy <- bvdTxFeePolicy <$> gsAdoptedBVData
+    case feePolicy of
```
Comment (volhovM) [inline src/Pos/Client/Txp/Util.hs]: Lambdacase, plus seems like a copy-paste of `prepareTxWithFee`.

File: src/Pos/Client/Txp/Util.hs [hunk 16]
```diff
@@ -206,60 +261,208 @@ prepareInpsOuts utxo addrs outputs = do
...
+    either throwError pure .
+    bimap invalidFee TxFee .
+    integerToCoin .
+    ceiling .
+    calculateTxSizeLinear linearPolicy .
+    biSize @TxAux
+  where
+    invalidFee reason = TxError ("Invalid fee: " <> reason)
+
+createFakeTxFromRawTx :: TxRaw -> TxAux
```
Comment (volhovM) [inline src/Pos/Client/Txp/Util.hs]: Comment.

File: src/Pos/Client/Txp/Util.hs [hunk 17]
```diff
@@ -174,27 +221,35 @@ makeRedemptionTx rsk txInputs = makeAbstractTx mkWit (map ((), ) txInputs)
...
+        Just inputsNE -> do
+            let totalTxAmount = unsafeSumTxOuts $ map snd inputsNE
+                trInputs = map formTxInputs inputsNE
+                trOutputs = outputs
+                trRemaining = totalTxAmount `unsafeSubCoin` totalMoneyWithFee
+            pure TxRaw {..}
   where
-    totalMoney = unsafeIntegerToCoin $ sumCoins $ map (txOutValue . toaOut) outputs
-    allUnspent = M.toList $ filterUtxoByAddrs (toList addrs) utxo
+    unsafeSumTxOuts = unsafeIntegerToCoin . sumCoins . map (txOutValue . toaOut)
```
Comment (flyingleafe) [inline src/Pos/Client/Txp/Util.hs]: Because one cannot send _that_ much money in one transaction?

File: lwallet/Main.hs [hunk 18]
```diff
@@ -145,23 +145,24 @@ addTxSubmit mvar = modifySharedAtomic mvar (\(TxCount submitted failed sending)
 addTxFailed :: Mockable SharedAtomic m => SharedAtomicT m TxCount -> m ()
 addTxFailed mvar = modifySharedAtomic mvar (\(TxCount submitted failed sending) -> return (TxCount submitted (failed + 1) sending, ()))
 
-runCmd :: forall ssc ctx m. MonadWallet ssc ctx m => SendActions m -> Command -> CmdCtx -> m ()
+runCmd :: SendActions LightWalletMode -> Command -> CmdCtx -> LightWalletMode ()
```
Comment (gromakovsky) [inline lwallet/Main.hs]: 👍

File: lwallet/Main.hs [hunk 19]
```diff
@@ -145,23 +145,24 @@ addTxSubmit mvar = modifySharedAtomic mvar (\(TxCount submitted failed sending)
...
-        withSafeSigner (skeys !! idx) (pure emptyPassphrase) $ \mss ->
-        runEitherT $ do
-            ss <- mss `whenNothing` throwError "Invalid passphrase"
-            lift $ submitTx
-                (immediateConcurrentConversations sendActions na)
-                ss
-                (map (flip TxOutAux []) outputs)
+    let skey = skeys !! idx
+        curAddr = makePubKeyAddress $ encToPublic skey
+    etx <- withSafeSigner skey (pure emptyPassphrase) $ \mss -> runEitherT $ do
```
Comment (gromakovsky) [inline lwallet/Main.hs]: `runEitherT`, omg (I know it was here already).

File: src/Pos/Generator/Block/Payload.hs [hunk 20]
```diff
@@ -196,7 +197,7 @@ genTxPayload = do
                     splitCoins outputsN (unsafeIntegerToCoin inputsSum)
                 let txOuts = NE.fromList $ zipWith TxOut outputAddrs coins
                 let txOutAuxsPre = map (\o -> TxOutAux o []) txOuts
-                either (lift . throwM . BGFailedToCreate) pure =<<
+                either (lift . throwM . BGFailedToCreate . unTxError) pure =<<
```
Comment (gromakovsky) [inline src/Pos/Generator/Block/Payload.hs]: I think it's more correct to attach `TxError` to `BGFailedToCreate`. If later we extend `TxError`, we will want this information in block generator too.

File: src/Pos/Client/Txp/Util.hs [hunk 21]
```diff
@@ -126,25 +157,27 @@ overrideTxOutDistrBoot c oldDistr = do
...
+               throwTxError $
                sformat ("Can't spend "%build%" coins: amount is too small for boot "%
                         " era and can't be distributed among genStakeholders") c
           -- TODO CSL-1351 boot stakeholders' weights are not used
           pure $ genesisSplitBoot (HM.fromList $ map (,1) genStakeholders) c
 
 -- | Same as 'overrideTxOutDistrBoot' but changes 'TxOutputs' all at once
-overrideTxDistrBoot ::
-       (TxCreateMode ctx m) => TxOutputs -> ExceptT Text m TxOutputs
+overrideTxDistrBoot
```
Comment (gromakovsky) [inline src/Pos/Client/Txp/Util.hs]: I think ideally `overrideTxDistrBoot` functions shouldn't be exported, because for me the look like an internal detail of other tx creation functions. All tx creation functions should use it. And it's quite useless separately.

File: src/Pos/Client/Txp/Util.hs (hunk 17 above)
Comment (gromakovsky) [inline src/Pos/Client/Txp/Util.hs]: It's still not clear to me, because I don't know this part well. What if someone calls `createRedemptionTx` and passes a list of outputs with too large sum of coins? It's wrong to do it, but if it happens, it makes more sense to throw `TxError` than to use `error` (which is unpredictable and dangerous).

**Comment by pva701** [comment]
Padajji moe review plz

File: src/Pos/Client/Txp/Addresses.hs [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+module Pos.Client.Txp.Addresses
+       ( MonadAddresses (..)
+       ) where
+
+import           Universum
+
+import           Pos.Types (Address)
+
+-- | A class which have the method to generate a new address
+class Monad m => MonadAddresses m where
```
Comment (pva701) [inline src/Pos/Client/Txp/Addresses.hs]: :+1:

Comment (pva701) [inline src/Pos/Client/Txp/Addresses.hs]: I dreamt about such type class

File: src/Pos/Client/Txp/Util.hs [hunk 23]
```diff
@@ -174,27 +221,35 @@ makeRedemptionTx rsk txInputs = makeAbstractTx mkWit (map ((), ) txInputs)
...
+                trOutputs = outputs
+                trRemaining = totalTxAmount `unsafeSubCoin` totalMoneyWithFee
+            pure TxRaw {..}
   where
-    totalMoney = unsafeIntegerToCoin $ sumCoins $ map (txOutValue . toaOut) outputs
-    allUnspent = M.toList $ filterUtxoByAddrs (toList addrs) utxo
+    unsafeSumTxOuts = unsafeIntegerToCoin . sumCoins . map (txOutValue . toaOut)
+    totalMoney = unsafeSumTxOuts outputs
+    totalMoneyWithFee = totalMoney `unsafeAddCoin` fee
+    allUnspent = M.toList utxo
```
Comment (pva701) [inline src/Pos/Client/Txp/Util.hs]: I can't drop my comment on `pickInputs` line, so it's here.

It a little bit differs from `selectSrcAddress` (in the previous code): there was a case when `moneyLeft` is less than current input and we tried to find an input with the same amount to set `remain = 0` (and don't use an extra input). you can see it in the last if-branch of `selectSrcAddressesDo`

I would prefer to have this logic here. what do you think?

Comment (pva701) [inline src/Pos/Client/Txp/Util.hs]: and it's subjective, but IMHO a function w/ several named parameters looks better than tuple-state and access to them via `_1`, `_2`... 
because non-familar reader is forced to see what this tuple means every time when he sees `_1`, `_2`, `_3`

File: src/Pos/Client/Txp/Util.hs [hunk 24]
```diff
@@ -206,60 +261,215 @@ prepareInpsOuts utxo addrs outputs = do
...
+    -> Utxo
+    -> TxOutputs
+    -> m TxRaw
+stabilizeTxFee linearPolicy utxo outputs =
+    stabilizeTxFeeDo 5 (TxFee $ mkCoin 0)
+  where
+    stabilizeTxFeeDo :: Int -> TxFee -> m TxRaw
+    stabilizeTxFeeDo 0 _ = throwTxError "Couldn't stabilize tx fee after 5 attempts"
+    stabilizeTxFeeDo attempt expectedFee = do
+        txRaw <- prepareTxRaw utxo outputs expectedFee
```
Comment (pva701) [inline src/Pos/Client/Txp/Util.hs]: `stabilizeTxFeeDo` uses `prepareTxRaw`, but `prepareTxRaw` isn't in `Fee logic` section.
it's hard to review and hard to follow the flow of logic of code when you read and modify this code.

File: src/Pos/Client/Txp/Util.hs [hunk 25]
```diff
@@ -206,60 +261,215 @@ prepareInpsOuts utxo addrs outputs = do
...
+            | otherwise = Just $ TxOutAux (TxOut fakeAddr trRemaining) []
+        txOutsWithRem = maybe trOutputs (\remTx -> remTx :| toList trOutputs) fakeOutMB
+
+        -- We create fake signers instead of safe signers,
+        -- because safe signer requires passphrase
+        -- but we don't want to reveal our passphrase to compute fee.
+        -- Fee depends on size of tx in bytes, sign of a tx has the fixed size
+        -- so we can use arbitrary signer.
+        srcAddrs = NE.map (txOutAddress . fst) trInputs
+        (_, fakeSK) = deterministicKeyGen "patakbardaqskovoroda228pva1488kk"
```
Comment (pva701) [inline src/Pos/Client/Txp/Util.hs]: it's so cute you perpetuated creator of this code in this string <3

File: src/Pos/Client/Txp/Util.hs [hunk 26]
```diff
@@ -206,60 +261,215 @@ prepareInpsOuts utxo addrs outputs = do
...
+    => Utxo
+    -> TxOutputs
+    -> ExceptT TxError m TxRaw
+prepareTxWithFee utxo outputs = withLinearFeePolicy $ \linearPolicy ->
+    stabilizeTxFee linearPolicy utxo outputs
+
+-- | Compute, how much fees we should pay to send money to given
+-- outputs
+computeTxFee
+    :: TxDistrMode ctx m
```
Comment (pva701) [inline src/Pos/Client/Txp/Util.hs]: it's too general constraint for this fucntion. AFAIU it would be enough `MonadError`

Comment (pva701) [inline src/Pos/Client/Txp/Util.hs]: ah ok, I forgot we have to get current fee policy from `GState`

File: src/Pos/Client/Txp/Util.hs (hunk 24 above)
Comment (flyingleafe) [inline src/Pos/Client/Txp/Util.hs]: @pva701 but `prepareTxRaw` is _not_ fee logic, it's basic logic of choosing the inputs of transaction, and it's used even when no fees are supposed to be added (like in redemption txs). I doubt that adding it to `Fee logic` section makes sense.

File: src/Pos/Client/Txp/Util.hs (hunk 23 above)
Comment (flyingleafe) [inline src/Pos/Client/Txp/Util.hs]: @pva701 regarding logic with trying to find the _exact_ same output: is it really worth it? I doubt that cases when we have some output with _exactly_ the same amount of money as what is remained to send are common - I bet they are quite rare. We'll only win one transaction output in such rare cases, which is not much in terms of bytes.

If we _really_ want to optimize it properly, then we need to implement something smarter (like reduction to maxflow), because taking biggest output first is not optimal indeed.

Comment (pva701) [inline src/Pos/Client/Txp/Util.hs]: ok, I realized if we even add this logic it won't prevent all cases taking of extra input.
so in general case, we take a minimum number of inputs or one more input, it is not worth it.
lets zab'em

**Comment by flyingleafe** [comment]
It works; tested manually and via Daedalus acceptance tests.
(Daedalus acceptance tests don't pass though, but only because they consider outgoing transaction amount invalid, if it's displayed with added transaction fees. Issue is reported to frontend team)
